test/e2e/client_tls_test.go

```diff
@@ -0,0 +1,468 @@
+		t.Fatalf("failed to create configmap %q: %v", clientCertsConfigmapName, err)
+	}
+	defer func() {
+		if err := kclient.Delete(context.TODO(), clientCertsConfigmap); err != nil {
+			t.Fatalf("failed to delete configmap %q: %v", clientCertsConfigmapName, err)
+		}
+	}()
+
+	// Create a test client pod with the client certificates.  We will exec
+	// curl commands in this pod to perform the tests.
+	podName := "client-tls-test-client"
+	image := deployment.Spec.Template.Spec.Containers[0].Image
+	clientPod := buildExecPod(podName, clientCertsConfigmapName.Namespace, image)
+	clientPod.Spec.Volumes = []corev1.Volume{{
+		Name: "client-certificates",
+		VolumeSource: corev1.VolumeSource{
+			ConfigMap: &corev1.ConfigMapVolumeSource{
+				LocalObjectReference: corev1.LocalObjectReference{
+					Name: clientCertsConfigmapName.Name,
+				},
+			},
+		},
+	}}
+	clientPod.Spec.Containers[0].VolumeMounts = []corev1.VolumeMount{{
+		Name:      "client-certificates",
+		MountPath: "/tmp/tls",
+		ReadOnly:  true,
+	}}
+	clientPodName := types.NamespacedName{
+		Name:      clientPod.Name,
+		Namespace: clientPod.Namespace,
+	}
+	if err := kclient.Create(context.TODO(), clientPod); err != nil {
+		t.Fatalf("failed to create pod %q: %v", clientPodName, err)
+	}
+	defer func() {
+		if err := kclient.Delete(context.TODO(), clientPod); err != nil {
+			t.Fatalf("failed to delete pod %q: %v", clientPodName, err)
+		}
+	}()
+
+	err = wait.PollImmediate(2*time.Second, 3*time.Minute, func() (bool, error) {
+		if err := kclient.Get(context.TODO(), clientPodName, clientPod); err != nil {
+			t.Logf("failed to get client pod %q: %v", clientPodName, err)
+			return false, nil
+		}
+		for _, cond := range clientPod.Status.Conditions {
+			if cond.Type == corev1.PodReady {
+				return cond.Status == corev1.ConditionTrue, nil
+			}
+		}
+		return false, nil
+	})
+	if err != nil {
+		t.Fatalf("timed out waiting for pod %q to become ready: %v", clientPodName, err)
+	}
+
+	// We need a client-go client in order to execute commands in the client
+	// pod.
+	kubeConfig, err := config.GetConfig()
+	if err != nil {
+		t.Fatalf("failed to get kube config: %v", err)
+	}
+	cl, err := kubernetes.NewForConfig(kubeConfig)
+	if err != nil {
+		t.Fatalf("failed to create kube client: %v", err)
+	}
+
+	// curl execs a Curl command in the test client pod and returns an error
+	// value.  The Curl command uses the specified certificate from the
+	// client certificates configmap, sends a request for the canary route
+	// via the router's internal service, and returns an error if the Curl
+	// command fails or the HTTP response status code indicates an error.
+	curl := func(cert string) error {
+		req := cl.CoreV1().RESTClient().Post().Resource("pods").
+			Namespace(clientPod.Namespace).Name(clientPod.Name).
+			SubResource("exec").
+			Param("container", clientPod.Spec.Containers[0].Name)
+		cmd := []string{
+			"/bin/curl", "-k", "-v",
```

Comment:
Is it practical to make a programmable HTTP request without needing to curl from within a pod?

Reply:
I should have been more explicit; something that just does:

```go
resp, err := http.Get("http://example.com/")
```

but in the context of this test I don't know how practical that would be. Given that this works end-2-end I'd leave it, at least for now.